File: algo/54. Spiral Matrix.js
// my solution used too much extra space

/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    if (matrix === null || matrix.length == 0) {
        return [];
    }
    var deltaRow = [0, 1, 0, -1];
    var deltaCol = [1, 0, -1, 0];
    var directionIndex = 0;
    var visited = [];
    for (var i=0; i<matrix.length; i++) {
        var tempRow = [];
        for (var j=0; j<matrix[0].length; j++) {
            tempRow.push(0);
        }
        visited.push(tempRow);
    }
    var cursorRow = 0, cursorCol = 0;
    var res = [];
    while (res.length<matrix.length*matrix[0].length) {
        res.push(matrix[cursorRow][cursorCol]);
        visited[cursorRow][cursorCol] = 1;
        var newCursorRow = cursorRow+deltaRow[directionIndex];
        var newCursorCol = cursorCol+deltaCol[directionIndex];
        if (newCursorRow>=matrix.length || newCursorCol>=matrix[0].length || newCursorRow<0 || newCursorCol<0 || visited[newCursorRow][newCursorCol] ) {
            directionIndex = (directionIndex+1)%4;
        }
        cursorRow += deltaRow[directionIndex];
        cursorCol += deltaCol[directionIndex];
    }
    return res;
};

File: round3/54. Spiral Matrix.cs
public class Solution {
    public IList<int> SpiralOrder(int[][] matrix) {
        var res = new List<int>();
        
        int m = matrix.Length;
        int n = matrix[0].Length;
        
        int offsetX = 0, offsetY = 0;
        
        while (offsetX < (m+1)/2 && offsetY < (n+1)/2) {
            
            if (offsetY * 2 + 1 == n) {
                for (int i = offsetX; i < m - offsetX; i ++) {
                    res.Add(matrix[i][n-offsetY-1]);
                } 
            } else if (offsetX * 2 + 1 == m) {
                for (int j = offsetY; j < n - offsetY; j ++) {
                    res.Add(matrix[offsetY][j]);
                }
            } else {
                for (int j = offsetY; j < n - offsetY - 1; j ++) {
                    res.Add(matrix[offsetY][j]);
                }
                for (int i = offsetX; i < m - offsetX - 1; i ++) {
                    res.Add(matrix[i][n-offsetY-1]);
                }
                for (int j = n - offsetY - 1; j > offsetY; j --) {
                    res.Add(matrix[m-offsetX-1][j]);
                }
                for (int i = m - offsetX - 1; i > offsetX; i --) {
                    res.Add(matrix[i][offsetY]);
                }
            }            
            
            offsetX ++;
            offsetY ++;
        }
        
        return res;
    }
}
